### PR TITLE
[YUNIKORN-186] clean up events in k8shim

### PR DESCRIPTION
**Pull request by steinsgateted**
### What is this PR for?
state, events and the state machine are in the same location, like `application_state.go`,` task_state.go`, `node_state.go`, `scheduler_state.go` .

- Each handle function removes `event *fsm.Event` parameter, like: `postTaskAllocated(event *fsm.Event)` becomes
 `postTaskAllocated(allocUUID string, nodeID string)` .
 
- In `task.go`, both` beforeTaskAllocated` and `postTaskAllocated` are used `dispatcher.Dispatch(NewFailTaskEvent(task.applicationID, task.taskID, errorMessage))`
 This can be removed and reused the `handleFailEvent` method.
 
- Remove the `common/events/states.go` and put the states in individual files:
 `application_state.go`, `task_state.go`, `node_state.go`, `scheduler_state.go`.
 
- The original ` common/events/events.go` moved the EventType into individual files.
But the events interface remains, because if the interface is removed, `dispatch` and `cache` package will have import cycle.

- stateMachine moved to ` application_state.go, node_state.go, task_state.go, scheduler_state.go`.

- Move the original `application_event.go, node_event.go, task_event.go, scheduler_event.go` to` application_state.go, node_state.go, task_state.go, scheduler_state.go`.

- Move the original `application_event_test.go, task_event_test.go` to `application_state_test.go, task_state_test.go`.

### What type of PR is it?
* [ ] - Bug Fix
* [ ] - Improvement
* [ ] - Feature
* [ ] - Documentation
* [ ] - Hot Fix
* [x] - Refactoring

### Todos
* [ ] - Task

### What is the Jira issue?
https://issues.apache.org/jira/projects/YUNIKORN/issues/YUNIKORN-186

### How should this be tested?
run the original unit test

### Questions:
* [ ] - The licenses files need update.
* [ ] - There is breaking changes for older versions.
* [ ] - It needs documentation.

### CODE REVIEW

**Comment by craigcondit** [comment]
@steinsgateted this looks pretty good. Can you rebase against the latest master to resolve the conflicts? Additionally this PR ( https://github.com/apache/yunikorn-k8shim/pull/418 ) is likely to be merged soon, and I suspect this one will have conflicts with it as well. Looking forward to reviewing this one more closely.

File: pkg/cache/node_state.go [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+	New        string
+	Recovering string
+	Accepted   string
+	Healthy    string
+	Rejected   string
+	Draining   string
+}
+
+func SchedulerNodeStates() *NStates {
+	if storeSchedulerNodeStates == nil {
```
Comment (craigcondit) [inline pkg/cache/node_state.go:79]: As before, use sync.Once().

File: pkg/cache/nodes.go [hunk 2]
```diff
@@ -234,9 +234,9 @@ func hasReadyCondition(node *v1.Node) bool {
 	return false
 }
 
-func triggerEvent(node *SchedulerNode, currentState string, eventType events.SchedulerNodeEventType) {
+func triggerEvent(node *SchedulerNode, currentState string, eventType SchedulerNodeEventType) {
 	log.Logger().Info("scheduler node event ", zap.String("name", node.name),
-		zap.String("current state ", currentState), zap.String("transition to ", string(eventType)))
+		zap.String("current state ", currentState), zap.String("transition to ", eventType.String()))
```
Comment (craigcondit) [inline pkg/cache/nodes.go:239]: Should probably use zap.Stringer() here rather than explicit .String() call.

File: pkg/shim/scheduler_state.go [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+	Registered  string
+	Registering string
+	Recovering  string
+	Running     string
+	Draining    string
+	Stopped     string
+}
+
+func SchedulerStates() *SStates {
+	if storeScheduleStates == nil {
```
Comment (craigcondit) [inline pkg/shim/scheduler_state.go:96]: Use sync.Once()

File: pkg/cache/application.go [hunk 4]
```diff
@@ -721,14 +621,7 @@ func (app *Application) handleReleaseAppAllocationEvent(event *fsm.Event) {
...
 
-func (app *Application) handleReleaseAppAllocationAskEvent(event *fsm.Event) {
-	eventArgs := make([]string, 2)
-	if err := events.GetEventArgsAsStrings(eventArgs, event.Args); err != nil {
-		log.Logger().Error("fail to parse event arg", zap.Error(err))
-		return
-	}
-	taskID := eventArgs[0]
-	terminationTypeStr := eventArgs[1]
+func (app *Application) handleReleaseAppAllocationAskEvent(taskID string, terminationTypeStr string) {
```
Comment (craigcondit) [inline pkg/cache/application.go:624]: Can we use just `terminationType` here?

**Review by craigcondit — CHANGES_REQUESTED** [review]
@steinsgateted, this looks really good. I flagged a couple of simple things, mostly around initialization of globals -- rather than check for nil, use sync.Once() to initialize these structures the first time they are accessed. This is completely thread-safe.

File: pkg/cache/application_state.go [hunk 5]
```diff
@@ -0,0 +1,577 @@
...
+	Completed  string
+	Killing    string
+	Killed     string
+	Failing    string
+	Failed     string
+	Resuming   string
+}
+
+func ApplicationStates() *AStates {
+	if storeApplicationStates == nil {
```
[the remark this answers is not in the document]
Comment (steinsgateted) [reply pkg/cache/application_state.go]: Done. Thanks.

File: pkg/cache/node_state.go (hunk 1 above)
Comment (steinsgateted) [reply pkg/cache/node_state.go]: Done. Thanks.

File: pkg/cache/nodes.go (hunk 2 above)
Comment (steinsgateted) [reply pkg/cache/nodes.go]: Done. Thanks.

File: pkg/cache/task_state.go [hunk 6]
```diff
@@ -0,0 +1,432 @@
...
+	Killing    string
+	Killed     string
+	Failed     string
+	Completed  string
+	Any        []string // Any refers to all possible states
+	Terminated []string // Rejected, Killed, Failed, Completed
+}
+
+func TaskStates() *TStates {
+	if storeTaskStates == nil {
```
[the remark this answers is not in the document]
Comment (steinsgateted) [reply pkg/cache/task_state.go]: Done. Thanks.

File: pkg/shim/scheduler_state.go (hunk 3 above)
Comment (steinsgateted) [reply pkg/shim/scheduler_state.go]: Done. Thanks.

File: pkg/cache/application.go [hunk 7]
```diff
@@ -696,14 +603,7 @@ func (app *Application) handleFailApplicationEvent(event *fsm.Event) {
...
 
-func (app *Application) handleReleaseAppAllocationEvent(event *fsm.Event) {
-	eventArgs := make([]string, 2)
-	if err := events.GetEventArgsAsStrings(eventArgs, event.Args); err != nil {
-		log.Logger().Error("fail to parse event arg", zap.Error(err))
-		return
-	}
-	allocUUID := eventArgs[0]
-	terminationTypeStr := eventArgs[1]
+func (app *Application) handleReleaseAppAllocationEvent(allocUUID string, terminationTypeStr string) {
```
[the remark this answers is not in the document]
Comment (steinsgateted) [reply pkg/cache/application.go]: Done. Thanks.

File: pkg/cache/application.go (hunk 4 above)
Comment (steinsgateted) [reply pkg/cache/application.go]: Done. Thanks.

File: pkg/cache/context.go [hunk 8]
```diff
@@ -910,13 +910,13 @@ func (ctx *Context) TaskEventHandler() func(obj interface{}) {
 				log.Logger().Error("failed to handle application event")
 				return
 			}
-
 			if task.canHandle(event) {
 				if err := task.handle(event); err != nil {
 					log.Logger().Error("failed to handle task event",
 						zap.String("applicationID", task.applicationID),
 						zap.String("taskID", task.taskID),
 						zap.String("event", string(event.GetEvent())),
```
[the remark this answers is not in the document]
Comment (steinsgateted) [reply pkg/cache/context.go]: Done. Thanks.

File: pkg/shim/scheduler_state.go [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+	"sync"
+
+	"github.com/looplab/fsm"
+	"go.uber.org/zap"
+
+	"github.com/apache/yunikorn-k8shim/pkg/common/events"
+	"github.com/apache/yunikorn-k8shim/pkg/log"
+)
+
+var once sync.Once
```
Comment (craigcondit) [inline pkg/shim/scheduler_state.go:31]: Can we name this like the other Once instances? `once` is a very common name for a large package.

**Review by craigcondit — CHANGES_REQUESTED** [review]
Almost there, one minor nit.

File: pkg/shim/scheduler_state.go (hunk 9 above)
Comment (steinsgateted) [reply pkg/shim/scheduler_state.go:31]: The packages in the cache seem to have duplicate names.

Comment (craigcondit) [reply pkg/shim/scheduler_state.go:31]: And those should probably be changed as well. I like the pattern you used for the other Once() instances in this PR.

**Comment by steinsgateted** [comment]
@craigcondit Thanks for your review.
I would like some clarifications.
Need to open another jira to put all the `zap.String("", var.String())` to `zap.Stringer("", var)` ?
then I found 
```
func beforeHook(event TaskEventType) string {
    return fmt.Sprintf("before_%s", event.String())
}
```
.String() can be removed :
```
func beforeHook(event TaskEventType) string {
    return fmt.Sprintf("before_%s", event)
}
```

File: pkg/shim/scheduler_state.go (hunk 9 above)
Comment (steinsgateted) [reply pkg/shim/scheduler_state.go]: Done. Thanks.

**Review by craigcondit — APPROVED** [review]
+1 pending PR pre-commit checks. I will merge once these pass.

**Comment by steinsgateted** [comment]
@craigcondit Thank you for the explanation.

**Pull request closed without merging**